### PR TITLE
[CR] Add ShareLaTeX provider

### PR DESCRIPTION
**Pull request by danielneis**
Hello,

we're developing a provider for ShareLaTeX (www.sharelatex.com) and have copied the code from s3, made the needed changes, and got it working for serving the project list and folder strutcture of a project for OSF using the code in this pull request plus the code on this link for the osf side: https://github.com/CenterForOpenScience/osf.io/compare/develop...danielneis:sharelatex

We have some doubts about the automated tests.
Since the strutuct of each plugin is somewhat different we would like to get some advices before assuming an approach or another.

What are the minimum requirements for testing a new provider?

We look at several of the tests files and found that the figshare (https://github.com/rafaeldelucena/waterbutler/blob/add-sharelatex/tests/providers/figshare/test_provider.py) may be the best start point, is it ok to follow that code style for tests?

Kind regards,
Daniel

### CODE REVIEW

File: waterbutler/providers/sharelatex/provider.py [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+        url = self.build_url('project', self.project_id,'file', path.path)
+
+        resp = yield from self.make_request(
+            'GET',
+            url,
+            expects=(200, 206),
+            throws=exceptions.DownloadError,
+        )
+
+        return streams.ResponseStreamReader(resp, None, None, True)
```
Comment (danielneis) [inline waterbutler/providers/sharelatex/provider.py]: Hello, this is the case with ShareLatex current API that is being developed as part of this project:
https://github.com/sharelatex/web-sharelatex/pull/212

**Comment by danielneis** [comment]
Hello, @chrisseto 

we've done a lot of progress on tests and other things.
Would you please take another look at the current state of the code?

Kind regards,
Daniel

**Comment by felliott** [comment]
Hey @danielneis,

I'll be following up on this, but I have a couple of other providers to CR first.  I'll see if I can get to it next week.  Thanks for your effort on this!

Cheers,
Fitz

**Comment by danielneis** [comment]
hello @felliott 
would be nice to have your review on that =)
kind regards,
daniel

**Comment by felliott** [comment]
Hey @danielneis,

Sorry for the long delay, I'll start looking at this today.  IIRC, last time I looked at this, sharelatex.com didn't yet provide API keys and I needed to run a local copy of sharelatex.  Is that still the case?

Also, as a heads up, WB is undergoing a bunch of code churn at the moment.  We're trying to add throttling to our requests and there's a conversion to Python 3.5 that we want to get merged in the near future.  The 3.5 conversion will probably affect you the most, but it should mostly be a matter of syntax updates.  `yield from` will become `await`, and the `@asyncio.coroutine` decorator will be replaced by `async def`.   The current work is in the `feature/python-3.5` branch.  Please let me know if you have any questions!

Cheers,
@felliott

**Comment by felliott** [comment]
The code looks pretty good, but I'd like to run it locally before doing anything.  Can you ping me when you fix [this issue](https://github.com/CenterForOpenScience/osf.io/pull/4597#issuecomment-188962125) on the OSF side?  Thank you!

**Comment by felliott** [comment]
Does ShareLaTeX have a concept of folders?  And will this be a read-only provider (at least for now)?

**Comment by felliott** [comment]
Also, are there docs describing the structure of the API in your `web-sharelatex` API PR?  Something like http://waterbutler.readthedocs.org/en/latest/api.html?

**Comment by felliott** [comment]
@danielneis, I'm pausing this CR for now.  The WB code looks pretty good!  This and @kwierman's [Dryad provider](https://github.com/CenterForOpenScience/waterbutler/pull/130) might have some overlap since they're both currently read-only providers.  I'll be reviewing his tomorrow, so I might have some more feedback for you.  In the meantime, can you update this PR for the `feature/python-3.5` branch, and address my comments on the OSF PR and the web-sharelatex PR?  Again, I'm sorry it's taken so long to get to these.  I'll be on the Slack channel regularly if you have any questions.

Cheers,
@felliott

**Comment by danielneis** [comment]
@felliott  thanks for your review!
I've talked to the other guys on my team and tomorrow we'll work on the issues you raised.
kind regards,
 daniel

**Comment by felliott** [comment]
Closing this, as ShareLaTex lacks the pieces needed to integrate.  But the code shall live on in the [feature/sharelatex-integration](https://github.com/centerforopenscience/waterbutler/tree/feature/sharelatex-integration) branch.

**Pull request closed without merging**